server/shared/src/main/scala/org/http4s/server/middleware/ErrorHandling.scala

```diff
@@ -42,4 +44,53 @@ object ErrorHandling {
 
   def httpApp[F[_]: MonadThrow](httpApp: HttpApp[F]): HttpApp[F] =
     apply(httpApp)
+
+  object Custom {
+    def handleErrorWith[F[_]: MonadThrow, G[_], A](
+        pf: PartialFunction[Throwable, F[Response[G]]],
+        http: Kleisli[F, A, Response[G]],
+    ): Kleisli[F, A, Response[G]] =
+      Kleisli { (a: A) =>
+        http.run(a).recoverWith(pf)
+      }
+  }
+
+  object Recover {
+
+    def total[F[_]: MonadThrow, G[_], A](
+        http: Kleisli[F, Request[G], Response[G]]
+    ): Kleisli[F, Request[G], Response[G]] =
+      Kleisli { (a: Request[G]) =>
+        http.run(a).handleError(totalRecover(a.httpVersion))
+      }
+
+    def messageFailure[F[_]: MonadThrow, G[_], A](
+        http: Kleisli[F, Request[G], Response[G]]
+    ): Kleisli[F, Request[G], Response[G]] =
+      Kleisli { (a: Request[G]) =>
+        http.run(a).recover(messageFailureRecover(a.httpVersion))
+      }
+
+    def messageFailureRecover[F[_], G[_]](
```

Comment:
```suggestion
    def messageFailureRecover[G[_]](
```

Reply:
Also curious - should `messageFailureRecover` and `totalRecover` be a public API? They seem like quite specific things.